src/Scrutor/Decoration/ClosedTypeDecorationStrategy.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+
+namespace Scrutor.Decoration
+{
+    internal sealed class ClosedTypeDecorationStrategy : IDecorationStrategy
+    {
+        private readonly Type _serviceType;
+        private readonly Type? _decoratorType;
+        private readonly Func<object, IServiceProvider, object>? _decoratorFactory;
+
+        public ClosedTypeDecorationStrategy(Type serviceType, Type? decoratorType, Func<object, IServiceProvider, object>? decoratorFactory)
+        {
+            _serviceType = serviceType;
+            _decoratorType = decoratorType;
+            _decoratorFactory = decoratorFactory;
+        }
+
+        public Type ServiceType => _serviceType;
+
+        public bool CanDecorate(Type serviceType) => _serviceType == serviceType;
+
+        public Func<IServiceProvider, object> CreateDecorator(Type serviceType)
+        {
+            if (_decoratorType is not null)
+            {
+                return DecoratorInstanceFactory.Default(serviceType, _decoratorType);
+            }
+
+            if (_decoratorFactory is not null)
```

Comment:
Both these implementations check `_decoratorType` and `_decoratorFactory` (or throw `InvalidOperationException`). Would it be possible to pull this logic up a level? I also see that the `_decoratorFactory ` branch in both implementations is the same, which means in practice it's only the `_decoratorType` branch that's different.